cpp/src/arrow/csv/converter.cc

```diff
@@ -381,32 +383,98 @@ class NumericConverter : public ConcreteConverter {
 /////////////////////////////////////////////////////////////////////////
 // Concrete Converter for timestamps
 
+namespace {
+
+struct InlineISO8601 {
```

Comment:
Why not reuse `StringConverter<TimestampType>`?

Reply:
That class needs to know the TimeUnit but it is not known yet when this class is constructed